### PR TITLE
feat(eslint-plugin): added new rule typedef

### LINKED ISSUE #558: [typedef] What is ETA of the typedef rule?
I'm thinking of migrating from TSLint to ESLint. Reading the [roadmap](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/ROADMAP.md), I realized the [`typedef`](https://palantir.github.io/tslint/rules/typedef/
) rule is still "unimplemented".

Is there any plan to implement this rule? If yes, do we have any ETA?

Thank you!

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Adds the equivalent of TSLint's `typedef` rule.

Closes #558.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+```ts
+const [a]: number[] = [1];
+const [b]: [number] = [2];
+const [c, d]: [boolean, string] = [true, 'text'];
+```
+
+### arrowCallSignature
+
+Whether to enforce type annotations for return types of arrow functions.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: is this option worthwhile when we have `explicit-function-return-type`?
It seems like duplicated work?

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+['hello', 'world'].map((text): number => text.length);
+
+const mapper = {
+  map: (text): string => text + '...',
+};
+```
+
+### arrowParameter
+
+Whether to enforce type annotations for parameters of arrow functions.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: isn't this covered by the compiler option `noImplicitAny`?

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+['hello', 'world'].map((text: string) => text.length);
+
+const mapper = {
+  map: (text: string) => text + '...',
+};
+```
+
+### callSignature
+
+Whether to enforce type annotations for return types of functions and methods.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: as per `arrowCallSignature` option

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+  log(text: string): boolean {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+### memberVariableDeclaration
+
+Whether to enforce type annotations on member variables of classes.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: I was REALLY surprised to learn that typescript doesn't enforce this with `noImplicitAny`.

```TS
class Foo {
    prop; // DOES NOT ERROR!!!
}

const x = new Foo()
x.prop; // typeof === any
```

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+Examples of **correct** code with `{ "objectDestructuring": true }`:
+
+```ts
+const { length }: { length: number } = 'text';
+const [b, c]: [number, number] = Math.random() ? [1, 2] : [3, 4];
+```
+
+### parameter
+
+Whether to enforce type annotations for parameters of functions and methods.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: as per `arrowParameter`

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+### propertyDeclaration
+
+Whether to enforce type annotations for properties of interfaces and types.
+
+Examples of **incorrect** code with `{ "propertyDeclaration": true }`:
+
+```ts
+type Members = {
+  member;
+  [i: string]: ;
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: this is not valid typescript:

```ts
interface OtherMembers {
  otherMember;
  [i: number]: ;
//             ^ Type expected.
}
```
[repl](https://www.typescriptlang.org/play/#src=interface%20OtherMembers%20%7B%0A%20%20otherMember%3B%0A%20%20%5Bi%3A%20number%5D%3A%20%3B%0A%7D)

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+};
+
+interface OtherMembers {
+  otherMember;
+  [i: number]: ;
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: ditto

File: packages/eslint-plugin/docs/rules/typedef.md [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+### variableDeclaration
+
+Whether to enforce type annotations for variable declarations, excluding array and object destructuring.
+
+Examples of **incorrect** code with `{ "variableDeclaration": true }`:
+
+```ts
+const text = 'text';
+let initialText = 'text';
+let delayedText;
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/typedef.md]: SERIOUSLY? Why doesn't `noImplicitAny` doesn't catch this!?!

File: packages/eslint-plugin/src/rules/typedef.ts [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+  callSignature?: boolean;
+  memberVariableDeclaration?: boolean;
+  objectDestructuring?: boolean;
+  parameter?: boolean;
+  propertyDeclaration?: boolean;
+  variableDeclaration?: boolean;
+}
+type Option = keyof Options;
+
+const OPTION_ARRAY_DESTRUCTURING: Option = 'arrayDestructuring';
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/typedef.ts]: would this be better as an enum?
```ts
const enum OptionKeys {
    arrayDestructuring = 'arrayDestructuring',
    arrowCallSignature = 'arrowCallSignature',
    // ....
}

type Options = {
    [k in OptionKeys]?: boolean;
};
```
[repl](https://www.typescriptlang.org/play/#src=const%20enum%20OptionKeys%20%7B%0A%20%20%20%20arrayDestructuring%20%3D%20'arrayDestructuring'%2C%0A%20%20%20%20arrowCallSignature%20%3D%20'arrowCallSignature'%2C%0A%20%20%20%20%2F%2F%20....%0A%7D%0A%0Atype%20Options%20%3D%20%7B%0A%20%20%20%20%5Bk%20in%20OptionKeys%5D%3F%3A%20boolean%3B%0A%7D%3B%0Aconst%20x%3A%20Options%20%3D%20%7B%0A%20%20%20%20%5BOptionKeys.arrayDestructuring%5D%3A%20true%2C%0A%20%20%20%20%5BOptionKeys.arrowCallSignature%5D%3A%20true%0A%7D%0A)

File: packages/eslint-plugin/docs/rules/typedef.md (hunk 2 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/typedef.md]: Yeah. Some users use the rule to make it easier to onboard `noImplicitAny`, or as a stopgap until they get around to enabling it.

File: packages/eslint-plugin/src/rules/typedef.ts (hunk 9 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/typedef.ts]: Haha probably - I went off the old TSLint file's contents, which were written this way before string enum values.

File: packages/eslint-plugin/docs/rules/typedef.md (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/typedef.md]: 🤷‍♂ it existed in the old rule, so I ported it here. That brings up the question: why have `explicit-function-return-type` when `typedef` can do the same thing? Should `typedef` be split up into separate rules?

Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/typedef.md]: `explicit-function-return-type` goes one step further, with options for the user to control how much type inference they want to allow for their return types.

examples
- you can allow not including a return type if the function is in a place which has a function type with a return type `const x: () => number = () => 1`
- there's a PR right now to allow higher order functions (`() => () => () => (): number => 1`)

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/typedef.md]: Nifty. I'm a little confused - @bradzacher what would you prefer this PR do? I'm equally happy...

* ...changing this option in some way?
* ...removing this option?
* ...splitting `typedef` into granular rules with names like `explicit-*-type`?

File: packages/eslint-plugin/docs/rules/typedef.md (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/typedef.md]: Err, sorry, I don't follow - is that suggested text to add here?

**Comment by Shinigami92** [comment]
I really love the TSLint typedef rule and I cant migrate to eslint without the support of this rule. This seid, I have one proposal what can improve typedef. Let us decide which option throws a warning and which throws an error -> so with this CI/CD can fail only on some of the options.

**Comment by bradzacher** [comment]
@Shinigami92 - this is not possible - eslint only allows for a single error level for a rule. Either it's all a warning, or it's all an error.

**Comment by Shinigami92** [comment]
Ok :/ sad but ok

File: packages/eslint-plugin/docs/rules/typedef.md (hunk 3 above)
Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/typedef.md]: sorry - I meant refer to the discussion on `arrowCallSignature`.

File: packages/eslint-plugin/docs/rules/typedef.md (hunk 1 above)
Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/typedef.md]: I would probably look to removing the two options that conflict with `explicit-function-return-type` (`arrowCallSignature` and `callSignature`).

Just to help people migrate, might be good to add a note to the docs directing people to `explicit-function-return-type`.

I'm happy with the rest of the options being in here.
I don't see a need to spend the time / add the complexity of adding one rule per option unless someone decides there needs to be options for specific cases.

**Comment by Shinigami92** [comment]
TSLint currently introduces a new option `variable-declaration-ignore-function`
https://github.com/palantir/tslint/issues/2654

**Pull request merged**